### ISSUE TITLE
Recompile only on changes to output of stanc3 auto-formatter (feature request & design discussion)

### ISSUE THREAD

**Issue by mike-lawrence** [stan-dev/cmdstanr#423]
(While an issue already exists related to this, as well as a PR, I thought it would be cleaner to start a new issue to discuss things now that I've had time to work on it and have learned about the option space.)

At present, when an exe already exists, cmdstanr will recompile if either: 
1. it's a new session so we have no record of last compile time
2. it's an old session but the stanfile has a modification time that is later than our record of the last compile time

1 makes for annoyance when one is coming back to a project and starting a new session (but also for packages that bundle exes, as evidenced by the original issue that spawned all this), and 2 is an annoyance when the stanfile has changed but only in non-functional ways (touched, whitespace, comments, etc)

Granted the Stan compiler has become very fast nowadays, so the above annoyances are in turn diminishing in magnitude. However, I think it's similarly trivial to solve them in a way that has little chance of going awry. 

Specifically, I propose that any compilation of the model start with passing it through the stanc3 auto-formatter, yielding a character string that is then hashed to a digest. A record of this hash is then kept somewhere permanent across sessions, and only when this record is absent or changes do we recompile (and obvs "re"compile if there's no exe).

There was some back-and-forth on where to store this digest previously; at first I thought it made sense to append it to the name of the exe itself and still think that's a fairly bulletproof approach, but this had the downside of having some potential issues related to detecting and deleting outdated exes, plus aesthetic issues with exes having longer more obscure names. An alternative was suggested whereby the hash is stored in a file located in a directory, and where some other recent FRs suggest implementations that involve use of a `stan_scratch` folder, we could use this directory for storing this file too. I came up with a third option whereby the hash is written back to the stan file itself as a comment at the end, but it occurs to me that this does not cover the case where the user is writing code for a stan model as a string in an R script. So I think use of the `stan_scratch` folder would be best after all. 

Note that it has been reported that it would be desirable for the hash to recompute on changes to any `#include` files used in the stan file. To achieve this, I imagine it would be pretty straightforward to parse the stan file for includes, confirm they refer to files and hash the contents of those files too. But this might be best left for a separate later feature.

**Comment by mike-lawrence** [comment]
Oh! just thought I'd note that I just encountered a "bug" in the released version of how we do things. I asked for a model to compile, then immediately made a change to the stan file, saving before the compilation finished. I then tried to recompile and clearly the save time of the exe was still later than the save time of the stanfile, so it claimed that the exe was up to date. Low-probability edge case obviously, and one that an aware user can get around by re-saving or using the `force_recompile` argument, but I thought I'd mention it as something that wouldn't happen with this feature's proposed implementation.

**Comment by martinmodrak** [comment]
I did read that only after putting some work into #483 which tries to address a similar issue in a simpler manner. I appreciate there is a lot of work in this! I do have one quick technical feedback:
I think not recompiling the model on comments/whitespace changes is potentially confusing as it can result in error messages signalling a mismatched line number. So if that's a feature I think it should not be turned on by default to avoid tripping novice users.

**Comment by mike-lawrence** [comment]
In the interim, see how [aria](https://github.com/mike-lawrence/aria) uses a local for and hashing of the stanc outputs to avoid unnecessary recompilation.

**Comment by martinmodrak** [comment]
I think the description is inaccurate - cmdstanr compares the timestamps of the Stan file and the executable, this has nothing to do with the session. If you are using `write_stan_file` to create your Stan file, the function will avoid overwriting the file if it already exists and has the correct content and thus keep the timestamp and avoid recompilation. 

However, this fails if the Stan file was deleted in the meantime (e.g. because it was in the temp directory as in your case) and it also fails if a change is made to the file, but the change is then reverted (which changes timestamp) or if a whitespace only/comment only change is made.

The proposal is to store enough information to be able to detect the need for recompilation without relying on timestamps. This could be external (as Mike suggests) or it could be stored internally in the executable (as https://github.com/stan-dev/cmdstan/issues/1178) suggests. Storing in the executable is IMHO more preferable as you don't have to worry that the storage becomes out of sync with your executables, but the downside is that this requires changes to cmdstan.

**Comment by jgabry** [comment]
I'm doing some issue cleanup and I'm going to close this issue as it's been a while and I think this is in large part addressed (by #495) and anything unaddressed should be handled by the more reliable solution in cmdstan itself (https://github.com/stan-dev/cmdstan/issues/1178) not in cmdstanr
